zebra-script/src/lib.rs

```diff
@@ -56,85 +56,127 @@ impl From<zcash_script_error_t> for Error {
-
-    let ret = unsafe {
-        zcash_script::zcash_script_verify(
-            script_ptr,
-            script_len as u32,
-            amount,
-            tx_to_ptr,
-            tx_to_len as u32,
-            n_in,
-            #[cfg(not(windows))]
-            flags,
-            #[cfg(windows)]
-            flags.try_into().expect("why bindgen whyyy"),
-            consensus_branch_id,
-            &mut err,
-        )
-    };
+/// A preprocessed Transction which can be used to verify scripts within said
+/// Transaction.
+#[derive(Debug)]
+pub struct CachedFfiTransaction {
+    transaction: Arc<Transaction>,
+    precomputed: *mut std::ffi::c_void,
+}
 
-    if ret == 1 {
-        Ok(())
-    } else {
-        Err(Error::from(err))
+impl CachedFfiTransaction {
+    /// Construct a `PrecomputedTransaction` from a `Transaction`.
+    pub fn new(transaction: Arc<Transaction>) -> Self {
+        let tx_to = transaction
+            .zcash_serialize_to_vec()
+            .expect("serialization into a vec is infallible");
+
+        let tx_to_ptr = tx_to.as_ptr();
+        let tx_to_len = tx_to.len() as u32;
+        let mut err = 0;
+
+        let precomputed = unsafe {
+            zcash_script::zcash_script_new_precomputed_tx(tx_to_ptr, tx_to_len, &mut err)
+        };
+
+        Self {
+            transaction,
+            precomputed,
+        }
+    }
+
+    pub fn inputs(&self) -> &[transparent::Input] {
+        self.transaction.inputs()
+    }
+
+    /// Verify a script within a transaction given the corresponding
+    /// `transparent::Output` it is spending and the `ConsensusBranchId` of the block
+    /// containing the transaction.
+    ///
+    /// # Details
+    ///
+    /// The `input_index` corresponds to the index of the `TransparentInput` which in
+    /// `transaction` used to identify the `previous_output`.
+    pub fn is_valid(
+        &self,
+        branch_id: ConsensusBranchId,
+        (input_index, previous_output): (u32, transparent::Output),
+    ) -> Result<(), Error> {
+        let transparent::Output { value, lock_script } = previous_output;
+        let script_pub_key: &[u8] = lock_script.0.as_ref();
+        let n_in = input_index as _;
+
+        let script_ptr = script_pub_key.as_ptr();
+        let script_len = script_pub_key.len();
+
+        let amount = value.into();
+        let flags = zcash_script::zcash_script_SCRIPT_FLAGS_VERIFY_P2SH
+            | zcash_script::zcash_script_SCRIPT_FLAGS_VERIFY_CHECKLOCKTIMEVERIFY;
+
+        let consensus_branch_id = branch_id.into();
+
+        let mut err = 0;
+
+        let ret = unsafe {
+            zcash_script::zcash_script_verify_precomputed(
+                self.precomputed,
+                n_in,
+                script_ptr,
+                script_len as u32,
+                amount,
+                #[cfg(not(windows))]
+                flags,
+                #[cfg(windows)]
+                flags.try_into().expect("why bindgen whyyy"),
```

Comment:
This is a C enum issue. bindgen is correctly representing the implementation-defined behavior allowed by the C standard.

C enums don't have a guaranteed size:
> 6.7.2.2 Enumeration specifiers
> [...]
> Constraints
> The expression that defines the value of an enumeration constant shall be an integer constant expression that has a value representable as an int.
> [...]
> Each enumerated type shall be compatible with char, a signed integer type, or an unsigned integer type. The choice of type is implementation-defined, but shall be capable of representing the values of all the members of the enumeration.

http://www.open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf

If we want to avoid these kinds of conversions, we should avoid using enums in our C interfaces. Then bindgen will create Rust interfaces that don't require integer size conversions.

```suggestion
                flags.try_into().expect("zcash_script_SCRIPT_FLAGS_VERIFY_* enum values fit in a c_uint"),
```